### PR TITLE
Issue #680 "Partitioning doesn't work for Wells"

### LINKED ISSUE #680: Partitioning doesn't work for Wells
In GitLab by @JoerivanEngelen on Dec 6, 2023, 16:45

Partitioning doesn't work for Wells, because the ``clip_by_grid`` method doesn't work for IPointData (Wells):

* The clipping method is based on the assumption that ``active`` is already the clipped grid, which it isn't. It appears to be a mask.
* There is no implementation for StructuredGrids

We apparently also lack serious testing for Wells in this regard.

Looking at the code, I think it should be relatively easy to use ``imod.select.points_values`` to throw wells outside grid, it also allows easy support of StructuredGrids.

This will potentially fix this as well: https://gitlab.com/deltares/imod/imod-python/-/issues/553

### PR DESCRIPTION
**Pull request by Manangka**
In GitLab by @JoerivanEngelen on Dec 11, 2023, 13:18

Closes #680, #700 and #553

This PR fixes the following:
- Change behaviour Well.mask method so it masks in 2D grids instead deepcopying
- clip_by_grid should always operate on 2d grids, correct tests for that
- refactor test_mf6_partitioning to use pytest cases instead of pytest parametrization
- fix erronous default value for ``out_of_bounds`` in ``point_values``
- fix bug where well could not be assigned to the first cell in an unstructured grid

### CODE REVIEW

**Comment by JoerivanEngelen** [comment]
Things I encountered:
We are inconsistent with accepted dimensionalities for the ``clip_by_grid`` and ``mask`` methods: Sometimes 3D, sometimes 2D. Probably good to pick that up in a separate issue.

**Comment by Manangka** [comment]
Could you Prepend "Issue #680" to the PR title? That way it will also end up in the git commit message

File: imod/mf6/utilities/clip.py [hunk 1]
```diff
@@ -54,9 +54,13 @@ def clip_by_grid(
     package: IPointDataPackage, active: xu.UgridDataArray
 ) -> IPointDataPackage:
     """Clip PointDataPackage outside unstructured grid."""
+
+    domain_slice = get_active_domain_slice(active)
+    active_clipped = active.isel(domain_slice, missing_dims="ignore")
```
Comment (Manangka) [inline imod/mf6/utilities/clip.py:59]: Why is this necessary?  Wouldn't this just return active?

File: imod/mf6/wel.py [hunk 2]
```diff
@@ -494,18 +492,32 @@ def regrid_like(self, target_grid: GridDataArray, *_) -> "Well":
         grid-agnostic, instead this method clips the package based on the grid
         exterior.
         """
-        return clip_by_grid(self, target_grid)
+        target_grid_2d = target_grid.isel(layer=0, drop=True, missing_dims="ignore")
+        return clip_by_grid(self, target_grid_2d)
 
-    def mask(self, _) -> Package:
+    def mask(self, domain: GridDataArray) -> "Well":
```
Comment (Manangka) [inline imod/mf6/wel.py:498]: If you add `from __future__ import annotations` then you can add the Well type hint as a regular hint instead of a string: `-> Well`

File: imod/mf6/wel.py [hunk 3]
```diff
@@ -494,18 +492,32 @@ def regrid_like(self, target_grid: GridDataArray, *_) -> "Well":
...
+        # Drop layer coordinate if present, otherwise a layer coordinate is assigned
+        # which causes conflicts downstream when assigning wells and deriving
+        # cellids.
+        domain = domain.isel(layer=0, drop=True, missing_dims="ignore").drop(
+            "layer", errors="ignore"
+        )
+        point_active = points_values(domain, x=self.x, y=self.y)
+
+        is_inside_exterior = point_active == 1
+        selection = self.dataset.loc[{"index": is_inside_exterior}]
```
Comment (Manangka) [inline imod/mf6/wel.py:515]: The above lines is the same as the clip_by_grid method. Can't you call that method to reduce code duplication?

File: imod/select/points.py [hunk 4]
```diff
@@ -243,7 +243,7 @@ def points_indices(da, out_of_bounds="raise", **points):
     return indices
 
 
-def points_values(da, out_of_bounds="error", **points):
+def points_values(da, out_of_bounds="raise", **points):
```
Comment (Manangka) [inline imod/select/points.py:246]: Why is this needed?

File: imod/tests/test_mf6/test_multimodel/test_mf6_partitioning_unstructured.py [hunk 5]
```diff
@@ -206,3 +206,63 @@ def test_partitioning_unstructured_with_vpt_cells(
...
+)
+def test_partitioning_unstructured_with_well(
+    tmp_path: Path, circle_model: Modflow6Simulation, partition_name: str
+):
+    simulation = circle_model
+    # increase the recharge to make the head gradient more pronounced
+    simulation["GWF_1"]["rch"]["rate"] *= 100
+
+    # Add well
+    well = imod.mf6.Well(
```
Comment (luitjansl) [inline imod/tests/test_mf6/test_multimodel/test_mf6_partitioning_unstructured.py:224]: why do you add just 1 well? Isn't that a bit meager for testing this functionality?

File: imod/mf6/utilities/clip.py [hunk 6]
```diff
@@ -54,9 +54,13 @@ def clip_by_grid(
     package: IPointDataPackage, active: xu.UgridDataArray
 ) -> IPointDataPackage:
     """Clip PointDataPackage outside unstructured grid."""
+
```
Comment (luitjansl) [inline imod/mf6/utilities/clip.py:57]: in many tests, "active" is taken as the top layer of idomain; and the top layer of idomain may sometimes contain some vertical passthrough (vpt) cells, even if the cells below it are active. Should wells coinciding with vpt cells stay or go? Or should we raise an exception if active contains vpt cells? Or should we make a function to infer active from idomain, and only assign vpt if all layers are vpt in idomain for a particular cell?

**Review by luitjansl — COMMENTED** [review]
looks good, some comments

File: imod/mf6/utilities/clip.py (hunk 1 above)
Comment (JoerivanEngelen) [reply imod/mf6/utilities/clip.py:59]: The way we select data in the partitioning is with a 2D grid of inactive and active cells. The line:

```is_inside_exterior = active_clipped.grid.locate_points(points) != -1```

only checks if wells are located within the grid exterior. So inactive cells have to be removed from the grid.

File: imod/tests/test_mf6/test_multimodel/test_mf6_partitioning_unstructured.py (hunk 5 above)
Comment (JoerivanEngelen) [reply imod/tests/test_mf6/test_multimodel/test_mf6_partitioning_unstructured.py:224]: Good point, added this. Immediately found a bug and fixed it! https://github.com/Deltares/imod-python/issues/700

File: imod/mf6/utilities/clip.py (hunk 6 above)
Comment (JoerivanEngelen) [reply imod/mf6/utilities/clip.py:57]: Good point. I think we need to decide what should be 3D and what should be 2D, I'll create a separate issue for this.

**Review by luitjansl — APPROVED** [review]
